private/buf/cmd/buf/command/push/push.go

```diff
@@ -347,6 +388,306 @@ func validateLabelFlagValues(flags *flags) error {
+	}
+	if flags.Create {
+		gitDefaultBranch, err := getGitDefaultBranch(ctx, runner, container, remotes, input)
+		if err != nil {
+			return nil, err
+		}
+		createModuleVisibility, err := bufmodule.ParseModuleVisibility(flags.CreateVisibility)
+		if err != nil {
+			return nil, err
+		}
+		gitMetadataUploadOptions = append(
+			gitMetadataUploadOptions,
+			bufmodule.UploadWithCreateIfNotExist(createModuleVisibility, gitDefaultBranch),
+		)
+	}
+	return gitMetadataUploadOptions, nil
+}
+
+func getGitRemotes(
+	ctx context.Context,
+	runner command.Runner,
+	input string,
+) ([]string, error) {
+	buffer := bytes.NewBuffer(nil)
+	if err := runner.Run(
+		ctx,
+		gitCommand,
+		command.RunWithArgs("remote"),
+		command.RunWithStdout(buffer),
+		command.RunWithStderr(buffer),
+		command.RunWithDir(input),
+	); err != nil {
+		return nil, err
+	}
+	return strings.Split(strings.TrimSpace(buffer.String()), "\n"), nil
+}
+
+// This returns an upload option for all Git metadata labels. We set labels for the following
+// Git matadata:
+//   - tags on current commit
+//   - current branch
+func getGitMetadataLabelsUploadOptions(
+	ctx context.Context,
+	runner command.Runner,
+	input string,
+) (bufmodule.UploadOption, error) {
+	tags, err := getGitTagsOnCurrentCommit(ctx, runner, input)
+	if err != nil {
+		return nil, err
+	}
+	branch, err := getCurrentGitBranch(ctx, runner, input)
+	if err != nil {
+		return nil, err
+	}
+	labels := append(tags, branch)
+	if len(labels) > 0 {
+		return bufmodule.UploadWithLabels(labels...), nil
+	}
+	return nil, nil
+}
+
+func getGitTagsOnCurrentCommit(
+	ctx context.Context,
+	runner command.Runner,
+	input string,
+) ([]string, error) {
+	buffer := bytes.NewBuffer(nil)
+	if err := runner.Run(
+		ctx,
+		gitCommand,
+		command.RunWithArgs("tag", "--points-at", "HEAD"),
+		command.RunWithStdout(buffer),
+		command.RunWithStderr(buffer),
+		command.RunWithDir(input),
+	); err != nil {
+		return nil, err
+	}
+	if len(buffer.Bytes()) > 0 {
+		return strings.Split(strings.TrimSpace(buffer.String()), "\n"), nil
+	}
+	return nil, nil
+}
+
+func getCurrentGitBranch(
+	ctx context.Context,
+	runner command.Runner,
+	input string,
+) (string, error) {
+	buffer := bytes.NewBuffer(nil)
+	if err := runner.Run(
+		ctx,
+		gitCommand,
+		command.RunWithArgs("branch", "--show-current"),
+		command.RunWithStdout(buffer),
+		command.RunWithStderr(buffer),
+		command.RunWithDir(input),
+	); err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(buffer.String()), nil
+}
+
+func getGitMetadataSourceControlURLUploadOption(
+	ctx context.Context,
+	runner command.Runner,
+	remotes []string,
+	input string,
+) (bufmodule.UploadOption, error) {
+	remoteToURL := map[string]string{}
+	for _, remote := range remotes {
+		buffer := bytes.NewBuffer(nil)
+		if err := runner.Run(
+			ctx,
+			gitCommand,
+			command.RunWithArgs("config", "--get", fmt.Sprintf("remote.%s.url", remote)),
```

Comment:
Yeah -- this is something I went back and forth on. And these values may be different. You could have `git config --get origin.remote.url` return the SSH vs. HTTPS. This is related to @saquibmian's comment below and also your other comments on empty URL's and also the SSH URL. Let me collate the comments below and I'll propose something.

Reply:
Consolidating URL conversation here: so I think we should make a best effort on URL. The approach I took is:
- check that the URL has a valid scheme, if not, print a warning and set no source control URL
- if it has a valid scheme, parse the hostname. if the hostname is one of known hostnames, then we construct a `https` url for the user with the commit sha. otherwise we return just the URL to the source control repository, since there is no way to ascertain the route to the specific commit.